write-rustdoc-hide-lines/src/formatter.rs

```diff
@@ -0,0 +1,84 @@
+use std::{
+    fs::{self, File},
+    io::{self, BufRead},
+    path::Path,
+};
+
+use crate::{
+    code_block_definition::CodeBlockDefinition, hidden_ranges::get_hidden_ranges,
+    utils::visit_dir_md_files,
+};
+
+pub fn run(dir: &Path) {
+    println!("Formatting folder: {:?}", dir);
+
+    let result = visit_dir_md_files(dir, &|entry| {
+        println!("{:?}", entry.path());
+
+        let file = File::open(entry.path())?;
+        let file_len = file.metadata().unwrap().len().try_into().unwrap();
+        let mut contents = String::with_capacity(file_len);
+        let mut is_inside_rust_code_block = false;
+        let mut rust_block: Vec<String> = vec![];
+
+        for line in io::BufReader::new(file).lines() {
```

Comment:
I tried doing what I suggested, but since this is parsing a file using doc comments it was pretty much impossible to write a doc test with a doc comment inside the doc test, but essentially I wanted this to be extracted so we could have a test that looked like this:
```rust
#[cfg(test)]
mod tests {
    use indoc::indoc;

    #[test]
    fn format_file_test() {
        let file = indoc! {r#"
            ```rust
            # test
            # test 2
            fn not_hidden() {

            }
            # test 3
            ```
        "#}
        .as_bytes();

        let contents = super::format_file(std::io::BufReader::new(file), file.len());

        assert_eq!(
            contents.unwrap(),
            indoc! {r#"
                ```rust,hide_lines=1-2 6
                # test
                # test 2
                fn not_hidden() {

                }
                # test 3
                ```
            "#}
        );
    }
}
```

Reply:
If you just extract the loop to a function called `format_file` that takes a `std::io::BufRead` you can add that test directly

Reply:
I've refactored it to a `format_file` function that accepts a lines iterator (see b0606b2).
And I've added tests for: add, modify & remove annotations (see 8b2993b).

But… I've a doubt now, I noticed that the formatter modifies all line endings. For example:

![image](https://user-images.githubusercontent.com/188612/181750423-e48e9c0d-dc25-40c8-ace8-5b4c7f2810f8.png)

![image](https://user-images.githubusercontent.com/188612/181753947-7050e9c1-003b-45ba-bbaa-bbbbaabc319a.png)

I think it's not bad behaviour but I'm not sure if it's good either. 😅 
It's doing more than it should, as in theory this should only change the Markdown `rust` code blocks.

cc @alice-i-cecile

Reply:
For now, I'm going to assume that that's the good behavior. I've updated the tests accordingly, see c94e00f.

But, if you think we should respect the line endings then we'll need to change the approach, maybe we can get the code block ranges using regex and be more surgical… ¯\_(ツ)_/¯

For reference, right only two files are affected by this:

- `/ecs/detecting-changes/_index.md`
- `/ecs/ecs-magic/_index.md`

A third file (`/ecs/entities-components/_index.md`) gets modified, but only the `rust` code blocks.

Reply:
Yeah, I won't block on this; that's the right behavior anyways.

Reply:
Yeah, we should already enforce consistent line endings anyway so I don't think it's a big issue